java/dagger/hilt/android/plugin/src/test/kotlin/BuildCacheTest.kt

```diff
@@ -0,0 +1,112 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import org.gradle.testkit.runner.TaskOutcome.FROM_CACHE
+import org.gradle.testkit.runner.TaskOutcome.SUCCESS
+import org.junit.Assert.assertEquals
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.junit.rules.TemporaryFolder
+import java.util.*
+import kotlin.random.Random
+
+// Test that verifies the hilt class transform does not break the Gradle's remote build cache.
+class BuildCacheTest {
+  @get:Rule
+  val gradleHomeFolder = TemporaryFolder()
+
+  @get:Rule
+  val firstProjectDir = TemporaryFolder()
+
+  lateinit var firstGradleRunner: GradleTestRunner
+
+  @get:Rule
+  val secondProjectDir = TemporaryFolder()
+
+  lateinit var secondGradleRunner: GradleTestRunner
+
+  private val testId = UUID.randomUUID().toString()
+
+  @Before
+  fun setup() {
+    firstGradleRunner = createGradleRunner(firstProjectDir)
+    secondGradleRunner = createGradleRunner(secondProjectDir)
+  }
+
+  private fun createGradleRunner(folder: TemporaryFolder): GradleTestRunner {
+    val gradleRunner = GradleTestRunner(folder)
+    gradleRunner.addDependencies(
+      "implementation 'androidx.appcompat:appcompat:1.1.0'",
+      "implementation 'com.google.dagger:hilt-android:LOCAL-SNAPSHOT'",
+      "annotationProcessor 'com.google.dagger:hilt-compiler:LOCAL-SNAPSHOT'",
+    )
+    gradleRunner.runAdditionalTasks("--build-cache")
+    gradleRunner.addSrc(
+      srcPath = "minimal/MyApp.java",
+      srcContent =
+      """
+        package minimal;
+        
+        import android.app.Application;
+
+        @dagger.hilt.android.HiltAndroidApp
+        public class MyApp extends Application {
+          // random id inserted into the code to ensure that the first is never a cache hit and the second 
+          // run always is
+          public static String RANDOM_ID = "$testId";
+        }
+        """.trimIndent()
+    )
+    gradleRunner.setAppClassName(".MyApp")
+    return gradleRunner
+  }
+
+  // Verifies that library B and library C injected classes are available in the root classpath.
+  @Test
+  fun test_buildCacheHitOnRelocatedProject() {
+    val firstResult = firstGradleRunner.build()
+    assertEquals(firstResult.getTask(":transformDebugClassesWithAsm").outcome, SUCCESS)
+
+    val secondResult = secondGradleRunner.build()
+    val cacheableTasks = listOf(
+      ":checkDebugAarMetadata",
+      ":checkDebugDuplicateClasses",
+      ":compileDebugJavaWithJavac",
+      ":compressDebugAssets",
+      ":extractDeepLinksDebug",
+      ":generateDebugBuildConfig",
+      ":generateDebugResValues",
+      ":javaPreCompileDebug",
```

Comment:
should `javaCompile` also be in here?

Reply:
At this point it doesn't come from cache in the second run, so would need to check why that is.

Reply:
Another strategy is to include a list of all tasks in the build like we do in the Andriod cache fix plugin: https://github.com/gradle/android-cache-fix-gradle-plugin/blob/master/src/test/groovy/org/gradle/android/CrossVersionOutcomeAndRelocationTest.groovy#L540

If any tasks are added to the build they must be explicitly added there.

Reply:
> should `javaCompile` also be in here?

Oh its actually there, it is just named `compileDebugJavaWithJavac` not `javaCompile`